src/os/bluestore/bluestore_types.h

```diff
@@ -289,25 +298,55 @@ struct bluestore_blob_t {
     assert(0 == "we should not get here");
   }
 
+  uint32_t get_csum_chunk_size() const {
+    return 1 << csum_chunk_order;
+  }
+
+  /// return chunk (i.e. min readable block) size for the blob
+  uint64_t get_chunk_size(uint64_t dev_block_size) {
+    return has_csum() ? MAX(dev_block_size, get_csum_chunk_size()) : dev_block_size;
+  }
   /// return true if the logical range has never been used
-  bool is_unused(uint64_t offset, uint64_t length) const {
-    return unused.contains(offset, length);
+  bool is_unused(uint64_t offset, uint64_t length, uint64_t chunk_size) const {
+    if (!has_unused()) {
+      return false;
+    }
+    uint64_t start = offset / chunk_size;
+    uint64_t end = ROUND_UP_TO(offset + length, chunk_size) / chunk_size;
+    assert( end <= unused.size());
+    auto i = start;
+    for(; i < end && unused[i]; i++);
+    return i >= end;
```

Comment:
This seems problematic: min_alloc_size / chunk_size could be almost anything, which means we could overflow the 32 bits at any time.  I see two options:

1. we fix the bit vector at 32 bits, and just the blob into 32 chunks unconditionally.  We'll probably be more or less granular than chunk_size, but as long as we underestimate the unused regions we'll be correct.  In practice most cases it will divide evenly and we'll be fine.  If they don't, we'll end up taking the wal path for some of the random writes, but that is not the end of the world.

2. We make the bitmap variably sized, and use the first few bits to encode how big it is.

I like #1.. less complex.  We might end up using 32 bits when 16 is enough.. but on the other hand we could even just say 16 bits is enough.  If the HDD default is 64k min_alloc_size and 4k block size, 16 would be fine for that case.